tests/unit/s2n_ktls_io_test.c

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+#include "tls/s2n_ktls.h"
+
+S2N_RESULT s2n_ktls_set_control_data(struct msghdr *msg, char *buf, size_t buf_size,
+        int cmsg_type, uint8_t record_type);
+S2N_RESULT s2n_ktls_get_control_data(struct msghdr *msg, int cmsg_type, uint8_t *record_type);
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* Test s2n_ktls_set_control_data and s2n_ktls_get_control_data */
+    {
+        /* Test: Safety */
+        {
+            char buf[100] = { 0 };
+            struct msghdr msg = { 0 };
+            EXPECT_ERROR_WITH_ERRNO(s2n_ktls_set_control_data(NULL, buf, sizeof(buf), 0, 0),
+                    S2N_ERR_NULL);
+            EXPECT_ERROR_WITH_ERRNO(s2n_ktls_set_control_data(&msg, NULL, sizeof(buf), 0, 0),
+                    S2N_ERR_NULL);
+            EXPECT_ERROR_WITH_ERRNO(s2n_ktls_set_control_data(&msg, buf, 0, 0, 0),
+                    S2N_ERR_NULL);
+
+            uint8_t record_type = 0;
+            EXPECT_ERROR_WITH_ERRNO(s2n_ktls_get_control_data(NULL, 0, &record_type),
+                    S2N_ERR_NULL);
+            EXPECT_ERROR_WITH_ERRNO(s2n_ktls_get_control_data(&msg, 0, NULL),
+                    S2N_ERR_NULL);
+        };
+
+        /* Test: s2n_ktls_set_control_data msg is parseable by s2n_ktls_get_control_data */
```

Comment:
Can you add some tests where you're testing s2n_ktls_set_control_data on its own? It seems like this function is doing more than just setting a record type, but that's all you have tested atm.

Reply:
Other than the record type, s2n_ktls_set_control_data is also setting some fields on the msghdr, but those fields are checked by s2n_ktls_get_control_data and only really relevant to s2n_ktls_get_control_data / the socket.

Is there any specific test you think is missing?

Reply:
I was thinking that there should be some tests for all of the CMSG macros that get called in that function. But going over it with @toidiu it seems like those are all constant expressions so there's not a lot of weird logic to test.